# Повний цикл NLP-проекту

## I. Перевірка фактів на достовірність

У межах цієї задачі ви побудуєте систему видобування фактів на правилах, а також інструменти для оцінювання цієї системи та фактів, які вона добуватиме.

### 1. Домен

Виберіть домен, для якого можна побудувати невелику базу даних на основі Wikipedia, DBPedia, IMDB тощо. Проаналізуйте домен і напишіть програму для побудови бази даних. Приклади доменів:
- актор і фільми, в яких він знімався
- письменник і книжки, які він написав
- музичний гурт і його учасники/концерти/альбоми з роками діяльності/випуску
- компанія і всі її CEO/CTO з роками діяльності
- людина і всі її місця роботи з часовими проміжками
- політик і політичні партії, в яких він брав участь
- винахідник та його винаходи

### 2. Видобування фактів

2.1. Напишіть програму, яка шукає статтю у Вікіпедії про сутність, що належить до вашого домена, та витягає текст цієї статті.

2.2. Напишіть програму, яка опрацьовує текст статті (саме текст, а не таблички, якщо такі є) та витягає з нього інформацію про ваш домен. Цю інформацію ви будете порівнювати зі сформованою базою даних.

### 3. Оцінювання результатів

Розробіть метрику, яка покаже, наскільки інформація, яку ви дістали зі статті, збігається з інформацією в вашій базі даних. Скільки пропущеної інформації з кожного боку? Чи є часткові збіги? (Наприклад, ім'я СЕО збігається лише частково або ім'я СЕО збігається, а роки діяльності різні.)

### Приклад

1. З бази каверів пісень дістаємо пісні та виконавців цих пісень з роками виконання. Наприклад, *Fever* з https://secondhandsongs.com/work/886/versions стане
```
1956,Little Willie John,Sandra Meade	
1957,Earl Grant,Kay Martin and Her Bodyguards,Ray Peterson
1958,Peggy Lee,Maureen Evans,Sallie Blair
1959,Norma Bengell,Sam Butera and The Witnesses,Frankie Avalon	
1960,Elvis Presley
...
1982,The Jam,Amanda Lear
1983,Marine Girls
1984,The Neville Brothers
...
```

2.1. Скрейпимо інформацію про пісню з Вікіпедії: https://en.wikipedia.org/wiki/Fever_(Little_Willie_John_song).

2.2. Пишемо набір правил (за допомогою частин мови, синтаксичних дерев, регекспів тощо), які витягають потрібну нам інформацію з тексту:

*Elvis Presley released a near identical version to Lee's two years following her cover, for his 1960 album, Elvis Is Back!...*  
*During their 1982 world tour, the British group The Jam covered the song as part of a medley with their own "Pity Poor Alfie"...*  
```
1960,Elvis Presley
1982,The Jam
```

3. Оцінюємо результат.


# Possible data sources

I decided to select domain based on information I would be able to acquire.
First of all, it would be great to get bunch of statements about specific domain that could be
a) easily verified using constructed KB
b) have ground truth labels
c) versatile formulation, so rules could be build to account for that

## [Open Trivia DB](https://opentdb.com)


First data source I considered appropriate. Using REST API I downloaded 1k sets of 50 random samples of True/False questions using script:
```bash
#!/usr/bin/env bash

rm input_raw.json
touch input_raw.json
echo "[" >> input_raw.json
for i in `seq 1 1000`;
  do
    curl "https://opentdb.com/api.php?amount=50&type=boolean" | jq .results | sed "1,1d" | sed '$d'  >> input_raw.json
    echo "," >> input_raw.json
    echo "."
  done
echo "]" >> input_raw.json
```
analysis below

In [1]:
import numpy as np
import pandas as pd
import gzip
import html
import re
import codecs
import spacy
import itertools
import hashlib
from IPython.display import clear_output
import sys

In [117]:
with gzip.open('input_raw.json.gz') as input_file:
    input_raw_df = pd.read_json(input_file, orient='records')

In [118]:
input_raw_df.head(5)

category correct_answer difficulty incorrect_answers  \
0            Science & Nature          False       easy            [True]   
1  Entertainment: Video Games           True     medium           [False]   
2        Entertainment: Music           True       easy           [False]   
3                   Geography          False     medium            [True]   
4                   Geography           True       easy           [False]   

                                            question     type  
0  Igneous rocks are formed by excessive heat and...  boolean  
1  Amazon acquired Twitch in August 2014 for $970...  boolean  
2  The music group Daft Punk got their name from ...  boolean  
3       The flag of South Africa features 7 colours.  boolean  
4  A group of islands is called an &#039;archipel...  boolean

In [119]:
input_unique = input_raw_df.drop_duplicates(subset=['question'])

In [73]:
input_unique.to_csv("open_trivia_unique.csv", index=False)

In [8]:
input_unique.shape

(492, 6)

After duplication removal, we have only 492 statements left :(

In [9]:
input_unique.groupby('category')['question'].agg('count')

category
Animals                                   16
Art                                        4
Celebrities                                4
Entertainment: Board Games                 8
Entertainment: Books                       5
Entertainment: Cartoon & Animations        4
Entertainment: Comics                      4
Entertainment: Film                       26
Entertainment: Japanese Anime & Manga     21
Entertainment: Music                      29
Entertainment: Musicals & Theatres         1
Entertainment: Television                 16
Entertainment: Video Games               106
General Knowledge                         50
Geography                                 35
History                                   37
Mythology                                  8
Politics                                  16
Science & Nature                          31
Science: Computers                        32
Science: Gadgets                           3
Science: Mathematics                      15
S

After going through questions manually and based on other datasets available, I consider Movie domain and using questions from `Entertainment: Television`, `Entertainment: Film`, `Entertainment: Cartoon & Animations`, `Entertainment: Video Games`, manually selected to the subset of those that deal with the fact about %X being in %Y movie / series / games

In [22]:
input_unique[input_unique['category'] == 'Sports']['question'].apply(lambda x: print(x) or x)

Peyton Manning retired after winning Super Bowl XLIX.
Roger Federer is a famous soccer player.
Wilt Chamberlain scored his infamous 100-point-game against the New York Knicks in 1962.
Soccer player Cristiano Ronaldo opened a museum dedicated to himself.
Tennis was once known as Racquetball.
In association football, or soccer, a corner kick is when the game restarts after someone scores a goal.
Skateboarding will be included in the 2020 Summer Olympics in Tokyo.
In Rugby League, performing a &quot;40-20&quot; is punished by a free kick for the opposing team.
Formula E is an auto racing series that uses hybrid electric race cars.
Manchester United won the 2013-14 English Premier League.
The Olympics tennis court is a giant green screen.


16      Peyton Manning retired after winning Super Bow...
178              Roger Federer is a famous soccer player.
191     Wilt Chamberlain scored his infamous 100-point...
260     Soccer player Cristiano Ronaldo opened a museu...
318                 Tennis was once known as Racquetball.
390     In association football, or soccer, a corner k...
617     Skateboarding will be included in the 2020 Sum...
740     In Rugby League, performing a &quot;40-20&quot...
937     Formula E is an auto racing series that uses h...
1403    Manchester United won the 2013-14 English Prem...
1482    The Olympics tennis court is a giant green scr...
Name: question, dtype: object

## [Film Trivia 30](https://www.sporcle.com/games/sekula32/film-trivia-30-true-statements)
Another dataset that deals with movies and scraped from page's iframe directly and can be used to infer %X in movie %Y relation

In [81]:
# taken from https://www.sporcle.com/games/sekula32/film-trivia-30-true-statements
film_trivia_30 = """
Al Pacino and Dustin Hoffman never appeared in the same film
Alfred Hitchcock’s Psycho was the first American film ever to show a flushing toilet
Bob Hoskins and Paul Sorvino directed Super Mario Bros
Clint Eastwood became oldest Best Actor nominee, after receiving nomination for his role in Million Dollar Baby
Daniel Day-Lewis is only actor to win three Best Actor Oscars
Darth Vader only has 12 minutes of screen time in the original Star Wars
Denzel Washington is first African American to win the Best Actor Oscar
Die Hard was Alan Rickman's feature film debut
E.T. and Poltergeist were supposed to be the same film
Jason Scott Lee, who portrays Bruce Lee in Dragon: The Bruce Lee Story, is not related to Bruce Lee
Jason Voorhees is the killer in the original Friday the 13th
Jennifer Lawrence is youngest Best Actress winner ever
John Wayne never died in any of his films
Korbin Dallas and Zorg, hero and villain of The Fifth Element, never met each other
Nicolas Cage turned down the role of Aragorn in The Lord of the Rings
None of Kurosawa's films won Best Foreign Language Film Oscar
Quentin Tarantino wrote a screenplay for Layer Cake
Stanley Kubrick has one Oscar, but not for Best Director
Sylvester Stallone starred in a porn film
The Godfather: Part II is only sequel that won Oscar for Best Picture
The Matrix marks only collaboration between Keanu Reeves and Laurence Fishburne
The nude portrait of Kate Winslet in Titanic was actually drawn by Leonardo DiCaprio
The rain in Singin' in the Rain was actually milk
The Shawshank Redemption, #1 on IMDb was the 51st highest grossing film in 94′
The Shining received two Razzie nominations
The voice of the velociraptors communicating in Jurassic Park are actually children screaming
The word 'Zombie' is never used in Night of the Living Dead
Three Oscar winners portrayed Batman on film
Tim Burton is the director of The Nightmare Before Christmas
While writing 'Good Will Hunting' Ben Affleck and Matt Damon included a scene of gay sex
"""
"OK"

'OK'

## [The Moview Dataset](https://www.kaggle.com/rounakbanik/the-movies-dataset/data)

Might be considered for the role of KB

In [23]:
#movies_metadata - keep id, imdb_id, original_title, title
#credits.csv - keep cast[gender/name/id], id
from zipfile import ZipFile
import ast 
with ZipFile('the-movies-dataset.zip', 'r') as myzip:
    with myzip.open('credits.csv') as input_file:
        credits_df = pd.read_csv(input_file, usecols=('id', 'cast'))
    with myzip.open('movies_metadata.csv') as input_file:
        metadata_df = pd.read_csv(input_file, usecols=('id', 'imdb_id', 'original_title', 'title'))


In [24]:
def unwrap_cast(cast):
    return [{'id': person['id'], 'gender': person['gender'], 'name': person['name']} for person in ast.literal_eval(cast)]

credits_df['cast'] = credits_df['cast'].apply(unwrap_cast)

In [33]:
credits_df.to_csv('movie_chars.csv', index = False)

In [37]:
credits_df['cast'][0]

[{'gender': 2, 'id': 31, 'name': 'Tom Hanks'},
 {'gender': 2, 'id': 12898, 'name': 'Tim Allen'},
 {'gender': 2, 'id': 7167, 'name': 'Don Rickles'},
 {'gender': 2, 'id': 12899, 'name': 'Jim Varney'},
 {'gender': 2, 'id': 12900, 'name': 'Wallace Shawn'},
 {'gender': 2, 'id': 7907, 'name': 'John Ratzenberger'},
 {'gender': 1, 'id': 8873, 'name': 'Annie Potts'},
 {'gender': 0, 'id': 1116442, 'name': 'John Morris'},
 {'gender': 2, 'id': 12901, 'name': 'Erik von Detten'},
 {'gender': 1, 'id': 12133, 'name': 'Laurie Metcalf'},
 {'gender': 2, 'id': 8655, 'name': 'R. Lee Ermey'},
 {'gender': 1, 'id': 12903, 'name': 'Sarah Freeman'},
 {'gender': 2, 'id': 37221, 'name': 'Penn Jillette'}]

In [70]:
metadata_df

imdb_id                     original_title  \
id                                                     
862     tt0114709                          Toy Story   
8844    tt0113497                            Jumanji   
15602   tt0113228                   Grumpier Old Men   
31357   tt0114885                  Waiting to Exhale   
11862   tt0113041        Father of the Bride Part II   
949     tt0113277                               Heat   
11860   tt0114319                            Sabrina   
45325   tt0112302                       Tom and Huck   
9091    tt0114576                       Sudden Death   
710     tt0113189                          GoldenEye   
9087    tt0112346             The American President   
12110   tt0112896        Dracula: Dead and Loving It   
21032   tt0112453                              Balto   
10858   tt0113987                              Nixon   
1408    tt0112760                   Cutthroat Island   
524     tt0112641                             Casino   
4584    tt0114388              Sense and Sensibility   
5       tt0113101                         Four Rooms   
9273    tt0112281     Ace Ventura: When Nature Calls   
11517   tt0113845                        Money Train   
8012    tt0113161                         Get Shorty   
1710    tt0112722                            Copycat   
9691    tt0112401                          Assassins   
12665   tt0114168                             Powder   
451     tt0113627                  Leaving Las Vegas   
16420   tt0114057                            Othello   
9263    tt0114011                       Now and Then   
17015   tt0114117                         Persuasion   
902     tt0112682         La Cité des Enfants Perdus   
37557   tt0115012                          摇啊摇，摇到外婆桥   
...           ...                                ...   
45527   tt1331329                    The Final Storm   
455661  tt6969946                     In a Heartbeat   
327237  tt3814486             Bloed, Zweet en Tranen   
84710   tt0036975                       Jungle Woman   
39562   tt0889600                  To Be Fat Like Me   
14008   tt0294425                        Cadet Kelly   
44330   tt0135690               Le tripot clandestin   
49279   tt0000359    L'Homme à la tête de caoutchouc   
44333   tt0135179                Les cartes vivantes   
49277   tt0135122           Les affiches en goguette   
49271   tt0127948            Le locataire diabolique   
44324   tt0135631               Le Roi du maquillage   
122036  tt0224286  Les Transmutations imperceptibles   
14885   tt0457437   Pooh's Heffalump Halloween Movie   
49280   tt0135453                  L'Homme orchestre   
106807  tt0135571     Nouvelles luttes extravagantes   
276895  tt3054038                        Deep Hearts   
404604  tt5690142                                Maa   
420346  tt4130180                  The Morning After   
67179   tt0069215         San Michele aveva un gallo   
84419   tt0038621                   House of Horrors   
390959  tt0265736          Shadow of the Blair Witch   
289923  tt0252966                The Burkittsville 7   
222848  tt0112613                    Caged Heat 3000   
30840   tt0102797                         Robin Hood   
439050  tt6209470                            رگ خواب   
111109  tt2028550                Siglo ng Pagluluwal   
67758   tt0303758                           Betrayal   
227506  tt0008536                Satana likuyushchiy   
461257  tt6980792                           Queerama   

                                   title  
id                                        
862                            Toy Story  
8844                             Jumanji  
15602                   Grumpier Old Men  
31357                  Waiting to Exhale  
11862        Father of the Bride Part II  
949                                 Heat  
11860                            Sabrina  
45325                       Tom and Huck  
9091                        Sudden Death  
710     

In [71]:
metadata_df.get_value('862', 'title')

'Toy Story'

In [34]:
metadata_df.to_csv('movie_list.csv', index = False)

## [200,000+ Jeopardy! Questions](https://data.world/sya/200000-jeopardy-questions)

In [2]:
nlp = spacy.load('en_core_web_md')

In [34]:
with codecs.open("JEOPARDY_CSV.csv", "r",encoding='utf-8', errors='ignore') as fdata:
    jeopardy = pd.read_csv(fdata)

Given a big qty of questions, idea is to filter out related questions based on
- words starred / acted / featured in question / category
- movie / cinema / screen / entities in category
Possibly, manually / using script fill blanks /  replace pronouns with answer, so  
```it starred Gregory Hines as a slumlord named Scrooge",A Christmas Carol```  
becomes  
```A Christmas Carol starred Gregory Hines as a slumlord named Scrooge```  
**Afterthought**: no need to search questions, good enough to find in categories only

In [43]:
nlp("A JIM CARREY FILM FESTIVAL".lower()).ents

(jim,)

In [45]:
jeopardy.head(5)

Show_Number  Air_Date      Round                         Category  Value  \
0         4680  12/31/04  Jeopardy!                          HISTORY  $200    
1         4680  12/31/04  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200    
2         4680  12/31/04  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200    
3         4680  12/31/04  Jeopardy!                 THE COMPANY LINE  $200    
4         4680  12/31/04  Jeopardy!              EPITAPHS & TRIBUTES  $200    

                                            Question      Answer  \
0  For the last 8 years of his life, Galileo was ...  Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe   
2  The city of Yuma in this state has a record av...     Arizona   
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...  John Adams   

  Category_entities  
0                ()  
1           ((10),)  
2                ()  
3                ()  
4                ()

In [44]:
jeopardy['Category_entities'] = jeopardy['Category'].apply(lambda x: nlp(x.lower()).ents)

In [46]:
jeopardy['Question_entities'] = jeopardy['Question'].apply(lambda x: nlp(x.lower()).ents)

In [53]:
jeopardy.to_csv('jeopardy_with_ents.csv', index = False)

In [54]:
jeopardy.head(5)

Show_Number  Air_Date      Round                         Category  Value  \
0         4680  12/31/04  Jeopardy!                          HISTORY  $200    
1         4680  12/31/04  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200    
2         4680  12/31/04  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200    
3         4680  12/31/04  Jeopardy!                 THE COMPANY LINE  $200    
4         4680  12/31/04  Jeopardy!              EPITAPHS & TRIBUTES  $200    

                                            Question      Answer  \
0  For the last 8 years of his life, Galileo was ...  Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe   
2  The city of Yuma in this state has a record av...     Arizona   
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...  John Adams   

  Category_entities         Question_entities  
0                ()  ((the, last, 8, years),)  
1           ((10),)   ((2), (1912), (6), (&))  
2                ()  ((yuma), (4,055, hours))  
3                ()     ((1963), (billionth))  
4                ()               ((second),)

In [3]:
# jeopardy = pd.read_csv('jeopardy_with_ents.csv')

In [4]:
jeopardy.head(5)

Show_Number  Air_Date      Round                         Category  Value  \
0         4680  12/31/04  Jeopardy!                          HISTORY  $200    
1         4680  12/31/04  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200    
2         4680  12/31/04  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200    
3         4680  12/31/04  Jeopardy!                 THE COMPANY LINE  $200    
4         4680  12/31/04  Jeopardy!              EPITAPHS & TRIBUTES  $200    

                                            Question      Answer  \
0  For the last 8 years of his life, Galileo was ...  Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe   
2  The city of Yuma in this state has a record av...     Arizona   
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...  John Adams   

  Category_entities    Question_entities  
0                ()  (the last 8 years,)  
1             (10,)      (2, 1912, 6, &)  
2                ()  (yuma, 4,055 hours)  
3                ()    (1963, billionth)  
4                ()            (second,)

In [6]:
jeopardy[['Category_entities', 'Question_entities', 'Answer', 'Category']]

Category_entities                               Question_entities  \
0                     ()                             (the last 8 years,)   
1                  (10,)                                 (2, 1912, 6, &)   
2                     ()                             (yuma, 4,055 hours)   
3                     ()                               (1963, billionth)   
4                     ()                                       (second,)   
5                     ()                                              ()   
6                     ()                                    (312, today)   
7                  (10,)                      (8, 30, birmingham, 2,306)   
8                     ()  (the winter of 1971-72, a record 1,122 inches)   
9                     ()                                              ()   
10                    ()                                              ()   
11                    ()                                        (first,)   
12                    ()                                         (1000,)   
13                 (10,)                          (1, &, syracuse &, 11)   
14                    ()                                (june 28, 1994,)   
15                    ()                              (1960, one minute)   
16                    ()                                              ()   
17                    ()                                              ()   
18                    ()                      (first, second, one, 1889)   
19                 (10,)                (10, the 1920s, mvp, 27 & ', 36)   
20                    ()                        (11 degrees, zero, 1935)   
21                    ()                                         (1898,)   
22                    ()                                         (1939,)   
23                    ()                                          (one,)   
24                    ()                                  (1885, indian)   
25                 (10,)                                     (5, kansas)   
26                    ()                                       (second,)   
27                    ()                                      (vermont,)   
28                    ()                                              ()   
29                    ()                                      (egg"</a,)   
...                  ...                                             ...   
216900                ()                                         (2006,)   
216901                ()                                           (14,)   
216902                ()                                              ()   
216903                ()                                (joseph conrad,)   
216904                ()                                         (1981,)   
216905                ()                                              ()   
216906                ()                                              ()   
216907                ()                               (4, 2, noon, & 3)   
216908                ()                                       (7-inch,)   
216909                ()                                         (1950,)   
216910                ()                                              ()   
216911                ()                                              ()   
216912                ()                                              ()   
216913                ()                                              ()   
216914                ()                                          (one,)   
216915                ()                                          (d.a,)   
216916                ()                                              ()   
216917                ()                                     (guyanese,)   
216918                ()                                    (18th, 2006)   
216919                ()                                         (1898,)   
216920                ()                               

In [39]:
category_filter = jeopardy.Category.str.contains("\s(character|act|play|movie|cinema|film|tv|screen|role|sitcom)", case=False)
jeopardy[category_filter]

/usr/local/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  """Entry point for launching an IPython kernel.


Show_Number  Air_Date             Round                    Category  \
118            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
124            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
130            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
136            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
142            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
179            3673   7/19/00         Jeopardy!                  1994 FILMS   
185            3673   7/19/00         Jeopardy!                  1994 FILMS   
191            3673   7/19/00         Jeopardy!                  1994 FILMS   
197            3673   7/19/00         Jeopardy!                  1994 FILMS   
203            3673   7/19/00         Jeopardy!                  1994 FILMS   
298            5690    5/8/09         Jeopardy!                 MOVIES & TV   
304            5690    5/8/09         Jeopardy!                 MOVIES & TV   
310            5690    5/8/09         Jeopardy!                 MOVIES & TV   
316            5690    5/8/09         Jeopardy!                 MOVIES & TV   
322            5690    5/8/09         Jeopardy!                 MOVIES & TV   
420            6037   12/7/10         Jeopardy!                 TIMELESS TV   
426            6037   12/7/10         Jeopardy!                 TIMELESS TV   
432            6037   12/7/10         Jeopardy!                 TIMELESS TV   
438            6037   12/7/10         Jeopardy!                 TIMELESS TV   
443            6037   12/7/10         Jeopardy!                 TIMELESS TV   
448            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
454            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
459            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
469            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
656            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
662            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
668            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
674            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
680            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
751            5851    2/8/10  Double Jeopardy!  THE STAR'S TV SHOW & MOVIE   
...             ...       ...               ...                         ...   
216296         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216302         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216308         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216314         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216320         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216327         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216333         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216339         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216345         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216351         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216357         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216363         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216369         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216375         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216380         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216477         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216483         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216489         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216495         5061   9/18/06  Double Jeopardy!           

In [40]:
extra_category_filter = jeopardy.Category.isin(["3", "10", "21", "1921", "1946", "1955", "1957", "1959", "1994", "2001",
"____ AND ____", "...AND MAN CREATED WOMAN",
"...OR BUST", "'40s POP CULTURE", "'49ers", "'60s FLICKS", "'60s MUSIC SCENE",
"'60s POP MUSIC", "A.M.", "AA"])
jeopardy[extra_category_filter]

Show_Number  Air_Date             Round                  Category  \
4948           3003   9/24/97  Double Jeopardy!                      1957   
4954           3003   9/24/97  Double Jeopardy!                      1957   
4960           3003   9/24/97  Double Jeopardy!                      1957   
4966           3003   9/24/97  Double Jeopardy!                      1957   
4972           3003   9/24/97  Double Jeopardy!                      1957   
15928          3660   6/30/00         Jeopardy!  ...AND MAN CREATED WOMAN   
15934          3660   6/30/00         Jeopardy!  ...AND MAN CREATED WOMAN   
15940          3660   6/30/00         Jeopardy!  ...AND MAN CREATED WOMAN   
15946          3660   6/30/00         Jeopardy!  ...AND MAN CREATED WOMAN   
15952          3660   6/30/00         Jeopardy!  ...AND MAN CREATED WOMAN   
23206          3540   1/14/00         Jeopardy!                ...OR BUST   
23212          3540   1/14/00         Jeopardy!                ...OR BUST   
23218          3540   1/14/00         Jeopardy!                ...OR BUST   
23224          3540   1/14/00         Jeopardy!                ...OR BUST   
23230          3540   1/14/00         Jeopardy!                ...OR BUST   
35318          5821  12/28/09         Jeopardy!                      2001   
35324          5821  12/28/09         Jeopardy!                      2001   
35330          5821  12/28/09         Jeopardy!                      2001   
35336          5821  12/28/09         Jeopardy!                      2001   
38813           618   4/22/87  Double Jeopardy!                      1946   
38819           618   4/22/87  Double Jeopardy!                      1946   
38834           618   4/22/87  Double Jeopardy!                      1946   
47460          2513    7/5/95         Jeopardy!                      1994   
47466          2513    7/5/95         Jeopardy!                      1994   
47472          2513    7/5/95         Jeopardy!                      1994   
47478          2513    7/5/95         Jeopardy!                      1994   
47484          2513    7/5/95         Jeopardy!                      1994   
66560          3195   6/19/98         Jeopardy!                        AA   
66566          3195   6/19/98         Jeopardy!                        AA   
66572          3195   6/19/98         Jeopardy!                        AA   
...             ...       ...               ...                       ...   
178210         4744   3/31/05         Jeopardy!                      1955   
178216         4744   3/31/05         Jeopardy!                      1955   
178222         4744   3/31/05         Jeopardy!                      1955   
178228         4744   3/31/05         Jeopardy!                      1955   
178234         4744   3/31/05         Jeopardy!                      1955   
182711         4466   1/26/04  Double Jeopardy!          '60s MUSIC SCENE   
182717         4466   1/26/04  Double Jeopardy!          '60s MUSIC SCENE   
182723         4466   1/26/04  Double Jeopardy!          '60s MUSIC SCENE   
182729         4466   1/26/04  Double Jeopardy!          '60s MUSIC SCENE   
182735         4466   1/26/04  Double Jeopardy!          '60s MUSIC SCENE   
185726         3313   1/20/99         Jeopardy!                      1957   
185732         3313   1/20/99         Jeopardy!                      1957   
185738         3313   1/20/99         Jeopardy!                      1957   
185744         3313   1/20/99         Jeopardy!                      1957   
185750         3313   1/20/99         Jeopardy!                      1957   
191652         5558   11/5/08  Double Jeopardy!                      2001   
191657         5558   11/5/08  Double Jeopardy!                      2001   
191662         5558   11/5/08  Double Jeopardy!                      2001   
191667         5558   11/5/08  Double Jeopardy!                      2001   
191672         5558   11/5/08  Double Jeopardy!                      2001   
1929

In [41]:
interesting_questions = jeopardy[category_filter | extra_category_filter]
interesting_questions

Show_Number  Air_Date             Round                    Category  \
118            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
124            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
130            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
136            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
142            3751  12/18/00         Jeopardy!           TV ACTORS & ROLES   
179            3673   7/19/00         Jeopardy!                  1994 FILMS   
185            3673   7/19/00         Jeopardy!                  1994 FILMS   
191            3673   7/19/00         Jeopardy!                  1994 FILMS   
197            3673   7/19/00         Jeopardy!                  1994 FILMS   
203            3673   7/19/00         Jeopardy!                  1994 FILMS   
298            5690    5/8/09         Jeopardy!                 MOVIES & TV   
304            5690    5/8/09         Jeopardy!                 MOVIES & TV   
310            5690    5/8/09         Jeopardy!                 MOVIES & TV   
316            5690    5/8/09         Jeopardy!                 MOVIES & TV   
322            5690    5/8/09         Jeopardy!                 MOVIES & TV   
420            6037   12/7/10         Jeopardy!                 TIMELESS TV   
426            6037   12/7/10         Jeopardy!                 TIMELESS TV   
432            6037   12/7/10         Jeopardy!                 TIMELESS TV   
438            6037   12/7/10         Jeopardy!                 TIMELESS TV   
443            6037   12/7/10         Jeopardy!                 TIMELESS TV   
448            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
454            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
459            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
469            6037   12/7/10  Double Jeopardy!    DEFENESTRATION IN CINEMA   
656            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
662            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
668            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
674            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
680            2735   6/21/96         Jeopardy!      ACTRESSES' FIRST FILMS   
751            5851    2/8/10  Double Jeopardy!  THE STAR'S TV SHOW & MOVIE   
...             ...       ...               ...                         ...   
216296         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216302         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216308         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216314         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216320         5236   5/21/07  Double Jeopardy!                  HOT MOVIES   
216327         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216333         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216339         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216345         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216351         3919   9/20/01         Jeopardy!          WHAT'S UP, TV DOC?   
216357         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216363         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216369         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216375         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216380         3919   9/20/01  Double Jeopardy!        ROBERT REDFORD ROLES   
216477         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216483         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216489         5061   9/18/06  Double Jeopardy!             SINGERS ON FILM   
216495         5061   9/18/06  Double Jeopardy!           

In [42]:
interesting_questions.to_csv('temp.csv')

In [84]:
cleaned_qs = interesting_questions.apply(lambda x: re.sub("(?<=[\s|^])(he|it|she|this|these|who|_+)(?=('s|\s|$|\"))", x.Answer, x.Question, count = 1), axis=1)

# [NELL](http://rtw.ml.cmu.edu/rtw/)

This one seems to be really noisy and have some related information, but not too suitable + issue with annotating other info :(

## Combining all questions

In [67]:
all_qs = list()

In [120]:
input_unique_list = input_unique[input_unique['category'].isin(["Entertainment: Television",
                                            "Entertainment: Film", 
                                            "Entertainment: Cartoon & Animations",
                                            "Entertainment: Video Games"])]['question'].apply(html.unescape).tolist()

In [113]:
trivia_list = film_trivia_30.split("\n")[1:-1]

In [114]:
japardy_list = cleaned_qs.tolist()

In [115]:
japardy_list

['Once Tommy Mullaney on "L.A. Law", John Spencer now plays White House chief of staff Leo McGarry on The West Wing series',
 'Barbra Streisand knows James Brolin played Lt. Col. Bill "Raider" Kelly on "Pensacola: Wings of Gold"',
 '(Hi, I\'m Wallace Langham)  I played Don Kirshner in VH1\'s TV movie about The Monkees quartet who sang "Daydream Believer"',
 'Teri Hatcher looked "shipshape" as one of the singing "mermaids" The Love Boat jumped on board this cruisin\' series in 1985',
 'On "Saturday Night Live", Will Ferrell\'s famous for playing Craig the Cheerleader, Janet Reno & moi',
 'Quentin Tarantino directed Pulp Fiction film & also had a bit role as Jimmy of Toluca Lake',
 'As mad bomber Howard Payne in Speed film, Dennis Hopper planted a bomb on an L.A. area transit bus',
 'Jean Vander Pyl, The Flintstones played Wilma in the original cartoon series, played Mrs. Feldspar in this movie adaptation',
 'Containing the hit "Can You Feel The Love Tonight", The Lion King was Disney\'s

In [125]:
all_lines = itertools.chain(input_unique_list, trivia_list, japardy_list)

In [126]:
good = 0
all = 0
for line in all_lines:
    clear_output()
    print("Select if this is about actor(s) in movie/show/series")
    print(line)
    selection = str(input('y/n/q')).lower().strip()
    all += 1
    if selection[0] == 'y':
        good += 1
        hex_string = hashlib.md5(line.encode('utf-8')).hexdigest()
        with open("data/output_{}.txt".format(hex_string), 'w+') as out:
            out.write(line)
        continue
    if selection[0] == 'n':
        continue
    if selection[0] == 'q':
        print('Done at {} / {}'.format(good, all))
        break

Select if this is about actor(s) in movie/show/series
He was born in St. John, New Brunswick, but his son Kiefer was born in London, England
y/n/qq
Done at 149 / 634


Going through first 634 samples I ended up annotating 149 sentences as aligning with what I need

## Further, I'm going to annotate sentences with following labels
- star_movie
- star_actor
- unrelated_movie
- unrelated_actor

Where unrelated_movie/unrelated_actor would be catch-all for movies/person that are not part of actor-in-movie pair that I'm looking ofr